### PR TITLE
Options to specify EXIF tags where keywords and "people in image" are found

### PR DESCRIPTION
**Pull request by galactictech**
- What's the current behaviour?
The current code only looks for keywords under IPTC.Keywords. 
The Mac Photos.app stores keywords under XMP.Subject, and also puts the names of identified people under XMP.PersonInImage.

- What does the PR change?
This PR adds six new options, and new %people album name pattern:
  - keyword-locations - an array of locations in which to look for keywords, such as IPTC.Keywords or XMP.Subject.
  - include-keywords - an array of keywords to be included (filtered in).
  - exclude-keywords - an array of keywords to be excluded (filtered out).
  - people-locations - an array of locations in which to look for people names, XMP.PersonInImage.
  - include-people - an array of people to be included (filtered in).
  - exclude-people - an array of people to be excluded (filtered out).
  - The people are referenced in album names using a new %people pattern (similar to the %keywords pattern).

The tests were updated, including some changes to the photo and file fixtures. 
I also pulled in the latest changes made to thumbsup/thumbsup today (12/30/2020).

- Does it introduce a breaking change for existing users?
It should not introduce any breaking changes.
  - The new logic for people only comes into play if the new people-locations option is specified  (since there is no default tag location for the people). I chose not to specify XMP.PersonInImage as a default because it doesn't appear to be a standard.
  - The new logic for keywords only comes into play if the new keyword-locations option is specified. If the new option is not specified then the old code is used (looking for keywords in the picasa files and then IPTC.Keywords). The new code to dynamically handle keyword-locations doesn't deal with picasa files (picasa is no longer supported by Google as far as I am aware), so I couldn't simply replace the old code with new code and a default location).

### CODE REVIEW

**Comment by galactictech** [comment]
I see codeclimate is showing three issues I can fix.

**Comment by galactictech** [comment]
I updated the code, and now there are no codeclimate issues.

**Comment by rprieto** [comment]
Hi @galactictech, thank you this is a great PR.
I have some quick feedback, but overall very keen on merging this.

I see three different parts: 
1. support for People, which is great!
2. support for including/excluding people and keywords, also great
3. support for customising where the fields are read from. This is where I have the most questions. For example:

Because `XMP:Keywords` and `XMP:PersonInImage` are very standard fields I'd actually like to use them by default. For example, Envato automatically merges `IPTC:Keywords` and `XMP:Keywords` when looking for keywords (https://help.author.envato.com/hc/en-us/articles/360000471443-Metadata-Requirements-for-Photos).

It then means:

- is it worth making it configurable?
- it we do, I'd like to solve that in a more generic way. For example what about captions? 
- specifying custom locations seems to disable Picasa handling. Maybe we need `Picasa:Keywords` type of notation.
- I was confused by the name "location" at first, since EXIF also has location information.

With that in mind are you OK removing the configuration aspect for now? We can tackle it separately.
Then I'm happy to merge straight away, and I'm planning to release a new version in the next few days.

**Comment by galactictech** [comment]
Thanks for the feedback.

Here are my thoughts and questions on your bullet points (keep in mind I am not super familiar with EXIF defails):
The code is setup to handle things being configurable, and as a developer and user I generally lean towards making things configurable. I also am using XMP:Subject for keywords (which is where they show up in the photos I export from Photos on MacOS), which you didn’t list as a standard. I am fine having defaults for the options like you listed, but can that default list include Subject? Or could the “location” options be available but hidden? If the options were removed, where in the code would be the preferred place to put in defaults so the current code that looks at opts can still work (to make it easy to add options later?
Can you provide more info on what you are thinking in terms of making things more generic ala “captions”? Are you thinking of “captions” as a generic concepts, with groups of captions (like keywords and people)? Conceptually that seems like it makes sense, but the code for handling them might be a bit more complicated. For example, in album-pattern.js would it mean looking for something like %captions[<groupname>] (e.g. %caption[keywords] and %captions[people]) instead of the hard-coded %keywords and %people?
To handle Picassa:Keywods, somewhere in metadata.js would need to be special case handling. I am guessing in the new findTags() method, in the loop it would have to check to see if the first part of the tag is Picassa.
Would you want the defaults to mimic the current behavior (of looking at Picassa first, and if tags are found there then not looking elsewhere?). Or is it ok to change the behavior a bit so Picassa is treated more like the others?
I picked “location” is the term just because that was my initial thought, not really for any special reason. Probably instead of people-locations the option should be people-fields or people-tags (and similar for keywords, or whatever if they options are made more general)? What do you think would be the preferred name?
My code in findTags (in metadata.js) is currently splitting the tag name on a period rather than a colon (i.e. it expects XMP.Subject rather than XMP:Subject). Based on your message it sounds like it should be splitting on colon (e.g. XMP:Subject), correct?

Thanks

Geoffrey

**Comment by rprieto** [comment]
Yes that's right, I'm suggesting to change the defaults with your suggestions:
- `Keywords = union(XMP.Subject, IPTC.Keywords)  # instead of just IPTC.Keywords`
- `People = XMP. PersonInImage`

As I type this, I realise a big typo in my previous comment, I meant to use `XMP.Subject` instead of `XMP.Keywords` which doesn't exist.

I also like to make things configurable if there's a need for it. That's why I'm keen to look at metadata-mapping configurability as a second step, assessing what the use cases are (especially if we already support all standard fields).

In summary I'm very happy to merge the following today:
- adding `XMP:Subject` values to the keywords array (union with IPTC.Keywords)
- new `people` metadata field based on `XMP. PersonInImage`
- new `%people` shorthand notation for mappers
- new flags `--include-keywords`, `--exclude-keywords`, `--include-people`, `--exclude-people`

**Comment by rprieto** [comment]
Last question, keeping in mind that some themes display the metadata next to each file.
When we use `--exclude-people John`, do you think
- it should remove John from `metadata.people` (current PR behaviour)
- it should leave John in the metadata, but not include John in `%people` albums. In this case the inclusion/excusion should happen in the AlbumPattern file.

**Comment by galactictech** [comment]
Ah. Yeah. It should be the second option, which means the filter logic needs to be removed from the metadata file and put into the album pattern file.

What is your timeline? I am not sure I’ll get to this today :)

**Comment by rprieto** [comment]
Sorry I meant "today" as in whenever you're ready :) Thank again for all this work and ideas.

**Comment by galactictech** [comment]
I made all the changes we talked about. See the commit messages and code. 
I forgot to include a commit message that it now handles both period (e.g. XMP.Subject) and colon (e.g. XMP:Subject) forms internally (thought the options were removed for now).

I don't know what to do about the package-lock.json conflict.

**Comment by rprieto** [comment]
Hi @galactictech. Thanks again for the PR! This is now available on master and will be included in the next release.

Hope you don't mind I simplified the mapping logic to keep it consistent with the rest. I also kept your original commit with `findInTags`, which will be handy in the future if we look at configurable mappings overall (not just for keywords but for all metadata).
